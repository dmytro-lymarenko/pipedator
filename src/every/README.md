- **Description**

Creates a validator which works like `.every()` method from an array:
it will fail when at least one value fails and it will succeed when all values do. **This validator requires a value to be an array**.

- **How to import**

```typescript
import { every } from 'pipedator';
// or
import { every } from 'pipedator/lib/every';

```
- **Signature**

```typescript
function every(validator: Validator, message?: string): Validator;
```
- **Parameters**

  - `validator` - a validator which is executed for each value.
  - `message` - (optional) custom message


- **Usage**

```typescript
every(string()).validate(['test']); // valid
every(string()).validate(['test', 'another']); // valid
every(string()).validate([]]); // valid
every(string()).validate(['text', 5]); // invalid because at least one value is not a string
every(string(), 'Custom message').validate([{}, 5]); // invalid with 'Custom message'
every(string()).validate(5); // throws an error. Value should be an ARRAY!
```