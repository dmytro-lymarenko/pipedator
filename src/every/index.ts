import { createValidator, findFirstError, Validator, getCurrentPath } from '../core';
import { pipe } from '../pipe';
import { array } from '../array';

/**
 * value should be an array
 * @param validator
 */
export function every<ValidValue = any>(validator: Validator) {
	return pipe(
		array(),
		createValidator<ValidValue>({
			// here value should be an array
			validate: (value, ctx) => {
				// find the first error among values
				const { error } = findFirstError(
					() => validator,
					i => value[i],
					i => ({
						...ctx,
						path: [...ctx.path, i.toString()],
					}),
					value.length
				);

				if (error) {
					return {
						path: getCurrentPath(ctx),
						message: 'All children should be valid',
						children: [error],
					};
				}

				return null;
			},
		})
	);
}
