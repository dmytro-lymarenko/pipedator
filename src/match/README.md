- **Description**

Creates a validator for matching a string against provided regexp.

- **How to import**

```typescript
import { match } from 'pipedator';
// or
import { match } from 'pipedator/lib/match';

```
- **Signature**

```typescript
function match<ValidValue = any>(regexp: RegExp, message?: string): Validator<ValidValue>;
```
- **Parameters**

  - `regexp` - provided regexp to test a string (value).
  - `message` - (optional) custom message.


- **Usage**

```typescript
match(/test/).validate('Some test have passed'); // valid
match(/test/).validate('Some text have be read'); // invalid
```