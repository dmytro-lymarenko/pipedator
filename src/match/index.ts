import { test } from '../test';

export function match(regexp: RegExp, message?: string) {
	return test(value => regexp.test(value), message || `Value should match regexp: ${regexp}`);
}
