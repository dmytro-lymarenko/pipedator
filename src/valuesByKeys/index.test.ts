import { valuesByKeys } from './index';
import { match } from '../match';
import { every } from '../every';

describe('valuesByKeys()', () => {
	it('should validate', () => {
		expect(valuesByKeys(['a', 'b'], every(match(/^test/))).validate({ a: 'testA', b: 'testB' })).toBeNull();
		expect(valuesByKeys(['a', 'b'], every(match(/^test/))).validate({ a: 'testA', b: 'testB', c: 5 })).toBeNull();
	});

	it('should fail', () => {
		expect(valuesByKeys(['a', 'b'], every(match(/^test/))).validate({ a: 'testA', b: 5 })).toMatchSnapshot();
	});

	it('should support custom message', () => {
		expect(valuesByKeys(['a', 'b'], every(match(/^test/)), 'Custom message').validate({ a: 'testA', b: 5 })).toMatchSnapshot();
	});
});
