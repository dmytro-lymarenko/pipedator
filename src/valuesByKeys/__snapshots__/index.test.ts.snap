// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valuesByKeys() should fail 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should match regexp: /^test/",
      "path": "b",
    },
  ],
  "message": "Values should be valid",
  "path": "",
}
`;

exports[`valuesByKeys() should support custom message 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should match regexp: /^test/",
      "path": "b",
    },
  ],
  "message": "Custom message",
  "path": "",
}
`;
